cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -306,75 +327,201 @@ public String toDisplayString(Subject subject) {
+        log.debug("challenge type: " + challengeType);
+        log.debug("credentials: " + credentials);
+
+        URI jwtIssuer = null;
+        try {
+            jwtIssuer = getJwtIssuer(credentials);
+            if (!jwtIssuer.normalize().equals(oidcClient.issuer.normalize())) {
+                log.debug("Token from untrusted issuer: " + jwtIssuer + " ignored");
+                jwtIssuer = null; // not the configured issuer, so don't use it
+            }
+        } catch (MalformedClaimException | InvalidJwtException | MalformedURLException e) {
+            log.debug("Cannot determine issuer from token", e);
+        }
+        List<Principal> validatedPrincipals = null;
+        if (jwtIssuer != null) {
+            try {
+                validatedPrincipals = validateWithPubKey(jwtIssuer, challengeType, credentials);
+            } catch (MalformedURLException | MalformedClaimException | InvalidJwtException e) {
+                log.debug("Cannot validate token with issuer public key", e);
+            }
+        }
+        if (validatedPrincipals == null) {
+            // public key validation did not work. Try the user info endpoint
+            try {
+                // makes the assumption that there's only one issuer (the configured one. This allows it to work with
+                // both JWT tokens (issuer specified in the token) and access tokens with no issuer specified
+                validatedPrincipals = validateWithUserInfo(raw, oidcClient.getUserInfoEndpoint());
+            } catch (ResourceAlreadyExistsException | ResourceNotFoundException | IOException | InterruptedException e) {
+                throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN,
+                        "Cannot validate token using user info endpoint", e);
+            }
+        }
+
+        s.getPrincipals().remove(raw);
+        for (Principal p : validatedPrincipals) {
+            s.getPrincipals().add(p);
+        }
+
+        // TODO - oidcScope not assigned yet
+        AuthorizationToken authToken = new AuthorizationToken(challengeType, credentials, oidcDomains, oidcScope);
+        s.getPublicCredentials().add(authToken);
+    }
+
+    private List<Principal> validateWithPubKey(URI jwtIssuer, String challengeType,String credentials)
+            throws MalformedURLException, InvalidJwtException, MalformedClaimException {
+        VerificationKeyResolver httpsJwksKeyResolver = getHttpsJwksVerificationKeyResolver(jwtIssuer, challengeType);
+        JwtConsumer jwtConsumer = new JwtConsumerBuilder()
+                .setRequireExpirationTime()
+                .setExpectedIssuers(true, jwtIssuer.toString())
+                .setVerificationKeyResolver(httpsJwksKeyResolver)
+                .setJwsAlgorithmConstraints(new AlgorithmConstraints(AlgorithmConstraints.ConstraintType.PERMIT, // whitelist
+                        AlgorithmIdentifiers.RSA_USING_SHA256)) // RS256 only
```

Comment:
Why just RS256? the config of SRC IAM says it supports a list of encryption and signing algorithms? Can't you tell from the JWT which was used? If not, does the coe have to just try all the advertised algs? ugh